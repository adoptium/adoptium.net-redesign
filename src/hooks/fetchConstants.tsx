import { useEffect, useState } from 'react';
import axios from 'axios';
import {capitalize} from '../util/capitalize'

const baseUrl = 'https://api.adoptium.net/v3';

export function fetchOses(isVisible: boolean): OperatingSystem[] {

    const [oses, setOses] = useState<OperatingSystem[]>([]);

    useEffect(() => {
        if (isVisible) {
        (async () => {
            const url = `${baseUrl}/types/operating_systems`;

            axios.get(url)
                .then(function (response) {
                    const newOses = response.data
                        .map(s => {
                            const o: OperatingSystem = { name: capitalize(s), value: s.toLowerCase() }
                            return o
                        })
                        .sort((o1, o2) => o1.name.localeCompare(o2.name))

                    setOses(newOses);
                })
                .catch(function (error) {
                    setOses([]);
                });
        })();
        }
    }, [isVisible]);

    return oses;
}

export function fetchArches(isVisible: boolean): Architecture[] {

    const [arches, setArches] = useState<Architecture[]>([]);

    useEffect(() => {
        if (isVisible) {
        (async () => {
            const url = `${baseUrl}/types/architectures`;

            axios.get(url)
                .then(function (response) {
                    const newArches = response.data
                        .map(s => {
                            const a: Architecture = { name: capitalize(s), value: s.toLowerCase() }
                            if(a.name === 'x32') a.name = 'x86'
                            return a;
                        })
                        .sort((a1, a2) => a1.name.localeCompare(a2.name))    

                    setArches(newArches);
                })
                .catch(function (error) {
                    setArches([]);
                });
        })();
        }
    }, [isVisible]);

    return arches;
}

export interface OperatingSystem {
    name: string
    value: string
}

export interface Architecture {
    name: string
    value: string
}
