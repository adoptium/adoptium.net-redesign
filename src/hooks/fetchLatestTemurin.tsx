import { useEffect, useState } from "react"
import { VersionMetaData } from "."
import axios from "axios"

const baseUrl = "https://api.adoptium.net/v3"

export function fetchLatestForOS(
  isVisible: boolean,
  version: number,
  os: string,
  arch: string,
): Binary | null {
  if (!os) {
    return null
  }

  const [binary, setBinary] = useState<Binary | null>(null)

  useEffect(() => {
    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/assets/feature_releases/${version}/ga?os=${os}&architecture=${arch}&image_type=jdk&jvm_impl=hotspot&page_size=1&vendor=eclipse`

        axios
          .get(url)
          .then(function (response) {
            const json: LatestTemurin = response.data[0]

            let binary_link = json.binaries[0].package.link
            let binary_checksum = json.binaries[0].package.checksum

            if (json.binaries[0].installer) {
              ;(binary_link = json.binaries[0].installer.link),
                (binary_checksum = json.binaries[0].installer.checksum)
            }

            const binary = {
              release_name: json.release_name,
              link: binary_link,
              checksum: binary_checksum,
            }
            setBinary(binary)
          })
          .catch(function (error) {
            setBinary(null)
          })
      })()
    }
  }, [isVisible, version, os, arch])

  return binary
}

export interface Binary {
  release_name: string
  link: URL
  checksum: string
}

export interface LatestTemurin {
  download_count: number
  id: string
  release_link: URL
  release_name: string
  release_type: string
  source?: {
    link: URL
    name: string
    size: number
  }
  timestamp: Date
  updated_at: Date
  vendor: string
  version_data: VersionMetaData
  binaries: [
    {
      architecture: string
      download_count: number
      heap_size: string
      image_type: string
      jvm_impl: string
      os: string
      package: {
        checksum: string
        checksum_link: URL
        download_count: number
        link: URL
        metadata_link: URL
        name: string
        signature_link: URL
        size: number
      }
      installer?: {
        checksum: string
        checksum_link: URL
        download_count: number
        link: URL
        metadata_link: URL
        name: string
        signature_link: URL
        size: number
      }
    },
  ]
}
