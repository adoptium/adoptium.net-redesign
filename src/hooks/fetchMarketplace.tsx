import { VersionMetaData } from "."
import axios from "axios"

const baseUrl = "https://marketplace-api.adoptium.net"

export async function getAllPkgsForVersion(
  version: number,
  os: string,
  architecture: string,
  package_type: string,
  vendors: string[] = [],
): Promise<MarketplaceRelease[] | null> {
  let params = "?"
  params += "feature_version=" + version

  if (os !== "any") {
    if (os === "alpine-linux") {
      params += "&os=alpine_linux"
    } else {
      params += "&os=" + os
    }
  }

  // loop through vendors and add to params
  if (vendors.length > 0) {
    vendors.forEach(vendor => {
      params += "&vendor=" + vendor
    })
  }

  if (architecture !== "any") {
    params += "&architecture=" + architecture
  }

  const url = baseUrl + "/v1/assets/latestForVendors" + params
  const data = await axios
    .get(url)
    .then(function (response) {
      return response.data
    })
    .catch(function (error) {
      return null
    })

  return data
}

export function getImageForDistribution(distribution: string) {
  switch (distribution) {
    case "temurin":
      return "/images/adoptium-logo.png"
    case "redhat":
      return "/images/redhat.svg"
    case "bisheng":
      return "/images/huawei.svg"
    case "zulu":
      return "/images/azul-logo.png"
    case "semeru":
      return "/images/ibm-logo.png"
    default:
      return `/images/${distribution}-logo.png`
  }
}

export interface MarketplaceRelease {
  release_name: string
  vendor: string
  binary: {
    os: string
    architecture: string
    image_type: string
    jvm_impl: string
    package: {
      name: string
      link: URL
      sha265sum: string
      sha256sum_link: URL
      signature_link: URL
    }
    installer?: [
      {
        name: string
        link: URL
        sha265sum: string
        sha256sum_link: URL
        signature_link: URL
      },
    ]
    timestamp: Date
    scm_ref: string
    openjdk_scm_ref: string
    distribution: string
    aqavit_results_link: URL
    tck_affidavit_link: URL
  }
  version: VersionMetaData
}
