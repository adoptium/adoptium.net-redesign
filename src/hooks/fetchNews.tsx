import { useEffect, useState } from "react"
import axios from "axios"

const baseUrl = "https://newsroom.eclipse.org/api"

export function fetchNewsItems(isVisible: boolean, page: number): News | null {
  const [news, setNews] = useState<NewsResponse>({ news: [], pagination: null })
  const [events, setEvents] = useState<EventItem[]>([])

  useEffect(() => {
    if (isVisible) {
      (async () => {
        setNews(await fetchLatestNews(page))
        setEvents(await fetchLatestEvents())
      })()
    }
  }, [isVisible, page])

  const newsAndEvents: News = {
    news: news,
    events: events,
  }

  return newsAndEvents
}

async function fetchLatestNews(page) {
  const url = new URL(`${baseUrl}/news`)
  url.searchParams.append("parameters[publish_to]", "adoptium")
  url.searchParams.append("page", page)
  url.searchParams.append("pagesize", "5")

  return axios
    .get(url.toString())
    .then(function (response) {
      return response.data
    })
    .catch(function (error) {
      return []
    })
}

async function fetchLatestEvents() {
  const url = new URL(`${baseUrl}/events`)
  url.searchParams.append("parameters[publish_to]", "adoptium")

  return axios
    .get(url.toString())
    .then(function (response) {
      return response.data.events
    })
    .catch(function (error) {
      return []
    })
}

export interface News {
  news: NewsResponse
  events: EventItem[]
}

export interface NewsResponse {
  news: NewsItem[]
  pagination: {
    page: number
    pagesize: number
    result_start: number
    result_end: number
    result_size: number
    total_result_size: number
  } | null
}

export interface NewsItem {
  id: string
  title: string
  body: string
  date: Date
  link: URL
}

export interface EventItem {
  id: string
  title: string
  infoLink: URL
  date: Date
}

export interface EventAPI {
  events: EventItem[]
}
