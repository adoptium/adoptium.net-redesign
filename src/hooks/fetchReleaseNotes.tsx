import { VersionMetaData } from "."
import { useEffect, useState } from "react"
import axios from "axios"

const baseUrl = "https://api.adoptium.net/v3/info/release_notes"

export function fetchReleaseNotesForVersion(
  isVisible: boolean,
  version: any,
  sortReleaseNotesByCallback?: Function,
): ReleaseNoteAPIResponse | null {
  if (!version) {
    return null
  }

  const [releaseNotes, setReleaseNotes] =
    useState<ReleaseNoteAPIResponse | null>(null)

  useEffect(() => {
    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/${version}`

        await axios
          .get(url.toString())
          .then(function (response) {
            let result = response.data
            if (sortReleaseNotesByCallback) sortReleaseNotesByCallback(result)
            setReleaseNotes(result)
          })
          .catch(function (error) {
            setReleaseNotes(null)
          })
      })()
    }
  }, [isVisible, version])

  return releaseNotes
}

export interface ReleaseNoteAPIResponse {
  id: string
  release_name: string
  release_notes: ReleaseNote[]
  vendor: string
  version_data: VersionMetaData
}

export interface ReleaseNote {
  id: string
  link: URL
  title: string
  backportOf?: string
  priority?: string
  component?: string
  subcomponent?: string
  type?: string
}
