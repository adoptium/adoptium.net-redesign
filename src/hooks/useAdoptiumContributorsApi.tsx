import { useEffect, useState } from "react"
import axios from "axios"

// List of repos that will be checked for contributions
const repositories = [
  "temurin-build",
  "ci-jenkins-pipelines",
  "infrastructure",
  "aqa-tests",
  "adoptium.net",
  "api.adoptium.net",
  "containers",
  "installer",
  "STF",
  "run-aqa",
  "TKG",
  "aqa-test-tools",
  "aqa-systemtest",
  "bumblebench",
  "jenkins-helper",
]

// List of users to exclude from random contributor
const excludedContributors = [
  "dependabot-preview[bot]",
  "dependabot[bot]",
  "github-actions[bot]",
  "eclipse-temurin-bot",
]

const randomValue = list => {
  return list[Math.floor(Math.random() * list.length)]
}

let repoToCheck = randomValue(repositories)

const CONTRIBUTORS_API_URI = `https://api.github.com/repos/adoptium/${repoToCheck}/contributors?per_page=1`

/**
 * Parses "Link" response header from node/contributors API
 * Returns page values for "next" and "last" API URLs
 * @param linkHeader
 */
function linkParser(linkHeader: string): {
  [keu: string]: {
    url: string
    page: number
  }
} {
  const regex = /<([^?]+\?per_page=1&[a-z]+=([\d]+))>;[\s]*rel="([a-z]+)"/g
  let array: RegExpExecArray | null = null
  const object = {}

  do {
    array = regex.exec(linkHeader)

    if (array) {
      object[array[3]] = {
        url: array[1],
        page: parseInt(array[2], 10),
      }
    }
  } while (array !== null)

  return object
}

/**
 * Retrieves max amount of contributors for Adoptium repos.
 * Returns array with random contributor index and max contributors found.
 */
async function getMaxContributors(): Promise<[number, number]> {
  // this call is used to know how many contributors there are in this repo
  // check the Link header to compute first and last
  const linksHeaderValue = await axios
    .get(CONTRIBUTORS_API_URI)
    .then(function (response) {
      return response.headers.get("Link")
    })
    .catch(function (error) {
      return undefined
    })

  if (linksHeaderValue) {
    const links = linkParser(linksHeaderValue)

    const randomPage =
      Math.floor(Math.random() * Math.floor(links.last.page)) + 1

    return [randomPage, links.last.page]
  }

  throw new Error("Failed to get amount if max contributors")
}

/**
 * Retrieves a contributor's object by it's index in API
 * only return 'type: User' to filter out Bot
 * @param randomPage
 */
async function getContributor(randomPage: number): Promise<Contributor | null> {
  const contributor = await axios
    .get(`${CONTRIBUTORS_API_URI}&page=${randomPage}`)
    .then(function (response) {
      return response.data[0] as ContributorApiResponse
    })
    .catch(function (error) {
      return undefined
    })

  if (!contributor) {
    return null
  }

  return {
    avatarUri: contributor.avatar_url,
    commitsListUri: `https://github.com/adoptium/${repoToCheck}/commits?author=${contributor.login}`,
    repo: repoToCheck,
    contributionsCount: contributor.contributions,
    login: contributor.login,
    profileUri: contributor.html_url,
  }
}

/**
 * Calls relative APIs and returns random contributor for Node.js main repo.
 * Trying to store cached data in localStorage in order to do less consequent requests
 */
async function fetchRandomContributor() {
  let maxContributors: number | null = null
  let fetchDate: number | null = null
  let needToRefetch = false
  const ONE_MONTH_MS = 2592000000

  const wlsMaxContributors = `${repoToCheck}_max_contributors`
  const wlsFetchDate = `${repoToCheck}_fetch_date`

  if (window.localStorage) {
    const maxContributorsStored =
      window.localStorage.getItem(wlsMaxContributors)
    const fetchDateStored = window.localStorage.getItem(wlsFetchDate)

    maxContributors = maxContributorsStored
      ? parseInt(maxContributorsStored, 10)
      : null
    fetchDate = fetchDateStored ? parseInt(fetchDateStored, 10) : null
  }

  if (fetchDate && Date.now() - fetchDate >= ONE_MONTH_MS) {
    needToRefetch = true
  }

  // If localStorage and data is less than 1 month old, fetch 1 time
  try {
    if (maxContributors && !needToRefetch) {
      return await getContributor(
        Math.floor(Math.random() * Math.floor(maxContributors)) + 1,
      )
    }
    const [randomPage, lastPage] = await getMaxContributors()

    let contributor = await getContributor(randomPage)
    while (excludedContributors.includes(contributor.login)) {
      contributor = await getContributor(randomPage)
    }

    if (window.localStorage) {
      window.localStorage.setItem(wlsFetchDate, String(Date.now()))
      window.localStorage.setItem(wlsMaxContributors, String(lastPage))
    }

    return contributor
  } catch {
    return null
  }
}

export function useAdoptiumContributorsApi(
  isVisible: boolean,
): Contributor | null {
  const [contributor, setContributor] = useState<Contributor | null>(null)
  useEffect(() => {
    if (isVisible) {
      (async () => {
        setContributor(await fetchRandomContributor())
      })()
    }
  }, [isVisible])

  return contributor
}

export interface ContributorApiResponse {
  login: string
  id: number
  url: string
  type: string
  contributions: number
  /* eslint-disable camelcase */
  node_id: string
  avatar_url: string
  gravatar_id: string
  html_url: string
  followers_url: string
  following_url: string
  gists_url: string
  starred_url: string
  subscriptions_url: string
  organizations_url: string
  repos_url: string
  events_url: string
  received_events_url: string
  site_admin: boolean
  /* eslint-enable camelcase */
}

export interface Contributor {
  avatarUri: string
  profileUri: string
  login: string
  contributionsCount: number
  commitsListUri: string
  repo: string
}
