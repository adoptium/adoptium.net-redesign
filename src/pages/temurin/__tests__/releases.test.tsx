import React from "react"
import { render } from "@testing-library/react"
import { afterEach, describe, expect, it, vi } from "vitest"
import { axe } from "vitest-axe"
import Releases, { Head } from "../releases"
import AxiosInstance from "axios"
import MockAdapter from "axios-mock-adapter"
import { mockOsesAPI, mockArchesAPI } from '../../../__fixtures__/hooks'

const mock = new MockAdapter(AxiosInstance)

vi.mock('../../../hooks/fetchConstants', () => {
  return {
    fetchOses: () => {
      return mockOsesAPI();
    },
    fetchArches: () => {
      return mockArchesAPI();
    }
  };
});

afterEach(() => {
  vi.clearAllMocks()
})

describe("Releases page", () => {
  it("renders correctly", () => {
    mock.onGet().reply(200, [], { pagecount: 0 })

    const { container } = render(<Releases />)
    // eslint-disable-next-line
    const pageContent = container.querySelector("main")

    expect(pageContent).toMatchSnapshot()
  })

  it("head renders correctly", () => {
    mock.onGet().reply(200, [], { pagecount: 0 })

    const { container } = render(<Head />)
    // eslint-disable-next-line
    const title = container.querySelector("title")
    expect(title?.textContent).toEqual("Latest Releases | Adoptium")
  })

  it("has no accessibility violations", async () => {
    mock.onGet().reply(200, [], { pagecount: 0 })

    const { container } = render(<Releases />)
    const results = await axe(container)
    // @ts-ignore
    expect(results).toHaveNoViolations()
  })
})
