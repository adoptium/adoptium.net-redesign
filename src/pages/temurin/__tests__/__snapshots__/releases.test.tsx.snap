// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Releases page > renders correctly 1`] = `
<main>
  <section
    class="py-5 text-center container"
  >
    <div
      class="row py-lg-5"
    >
      <div
        class="col-lg-10 col-md-8 mx-auto"
      >
        <h1
          class="fw-light"
        >
          Eclipse Temurinâ„¢ Latest Releases
        </h1>
        <div
          class="row align-items-center pt-3"
        >
          <div
            class="col-6 col-md-4"
          >
            <img
              alt="Temurin logo"
              class="img-fluid"
              src="/images/temurin-light.png"
              width="120"
            />
          </div>
          <div
            class="col-12 col-sm-6 col-md-8"
          >
            <p
              class="text-start"
            >
              Text
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row align-items-center pt-3"
    >
      <p
        class="text-center"
      >
        Text
      </p>
    </div>
    <div
      class="input-group mb-5 row g-2"
    >
      <div
        class="input-group-prepend flex-colunm col-12 col-md-3"
      >
        <label
          class="px-2 fw-bold"
          for="os"
        >
          Text
        </label>
        <select
          aria-label="OS Filter"
          class="form-select form-select-sm"
          data-testid="os-filter"
          id="os-filter"
        >
          <option
            value="any"
          >
            Text
          </option>
          <option
            value="aix"
          >
            AIX
          </option>
          <option
            value="alpine-linux"
          >
            Alpine Linux
          </option>
          <option
            value="linux"
          >
            Linux
          </option>
          <option
            value="mac"
          >
            macOS
          </option>
          <option
            value="solaris"
          >
            Solaris
          </option>
          <option
            value="windows"
          >
            Windows
          </option>
        </select>
      </div>
      <div
        class="input-group-prepend flex-colunm col-12 col-md-3"
      >
        <label
          class="px-2 fw-bold"
          for="arch"
        >
          Text
        </label>
        <select
          aria-label="Architecture Filter"
          class="form-select form-select-sm"
          data-testid="arch-filter"
          id="arch-filter"
        >
          <option
            value="any"
          >
            Text
          </option>
          <option
            value="aarch64"
          >
            aarch64
          </option>
          <option
            value="arm"
          >
            arm
          </option>
          <option
            value="ppc64"
          >
            ppc64
          </option>
          <option
            value="ppc64le"
          >
            ppc64le
          </option>
          <option
            value="s390x"
          >
            s390x
          </option>
          <option
            value="sparcv9"
          >
            sparcv9
          </option>
          <option
            value="x64"
          >
            x64
          </option>
          <option
            value="x86"
          >
            x86
          </option>
        </select>
      </div>
      <div
        class="input-group-prepend flex-colunm col-12 col-md-3"
      >
        <label
          class="px-2 fw-bold"
          for="package-type"
        >
          Text
        </label>
        <select
          aria-label="Package Type Filter"
          class="form-select form-select-sm"
          data-testid="package-type-filter"
          id="package-type-filter"
        >
          <option
            value="any"
          >
            Text
          </option>
          <option
            value="jdk"
          >
            JDK
          </option>
          <option
            value="jre"
          >
            JRE
          </option>
        </select>
      </div>
      <div
        class="input-group-prepend flex-colunm col-12 col-md-3"
      >
        <label
          class="px-2 fw-bold"
          for="version"
        >
          Text
        </label>
        <select
          aria-label="Version Filter"
          class="form-select form-select-sm"
          data-testid="version-filter"
          id="version-filter"
        >
          <option
            value="1"
          >
            1 - LTS
          </option>
          <option
            value="2"
          >
            2
          </option>
        </select>
      </div>
    </div>
    <table
      class="table table-bordered releases-table"
      id="download-table"
      style="border-spacing: 0 10px; border-collapse: separate;"
    >
      <tbody
        class="table-light"
      >
        <tr />
      </tbody>
    </table>
    <div
      aria-hidden="true"
      aria-labelledby="checksumModalLabel"
      class="modal fade"
      id="checksumModal"
      style="z-index: 10000;"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-lg"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
              id="checksumModalLabel"
            >
              Checksum (SHA256)
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="modal"
              type="button"
            />
          </div>
          <div
            class="modal-body text-start"
          >
            <p>
              The checksum is used to ensure the file has not been corrupted during download. You should compare the checksum of the file you received with the value below to ensure the file you received is complete and unmodified.
            </p>
            <p>
              On Windows use
               
              <code>
                certUtil -hashfile 
                <var>
                  file
                </var>
                 SHA256
              </code>
              , on Linux use
               
              <code>
                sha256sum 
                <var>
                  file
                </var>
              </code>
              , and on macOS use
               
              <code>
                shasum -a 256 
                <var>
                  file
                </var>
              </code>
              .
            </p>
            <input
              readonly=""
              style="width: 100%;"
              value=""
            />
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Copy
            </button>
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row align-items-center pt-3"
    >
      <p
        class="text-center"
      >
        Text
      </p>
      <div
        class="btn-group"
      >
        <a
          class="btn btn-primary m-3"
          href="/temurin/archive"
        >
          Text
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-Width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
            />
          </svg>
        </a>
        <a
          class="btn btn-secondary m-3"
          href="/temurin/nightly"
        >
          Text
           
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-Width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm-28.9 143.6l75.5 72.4H120c-13.3 0-24 10.7-24 24v16c0 13.3 10.7 24 24 24h182.6l-75.5 72.4c-9.7 9.3-9.9 24.8-.4 34.3l11 10.9c9.4 9.4 24.6 9.4 33.9 0L404.3 273c9.4-9.4 9.4-24.6 0-33.9L271.6 106.3c-9.4-9.4-24.6-9.4-33.9 0l-11 10.9c-9.5 9.6-9.3 25.1.4 34.4z"
            />
          </svg>
        </a>
      </div>
    </div>
  </section>
</main>
`;
